src/actors/symbolication.rs

```diff
@@ -97,37 +98,217 @@ type ComputationChannel = Shared<oneshot::Receiver<(Instant, SymbolicationRespon
+                                ObjectFileStatus::from(&err)
+                            }
+                        };
+                        CfiModule {
+                            features: cfi_cache.features(),
+                            cfi_status,
+                            cfi_path: Some(cfi_cache.path().to_owned()),
+                            scanned: false,
+                        }
+                    }
+                    Err(err) => {
+                        log::debug!("Error while fetching cficache: {}", LogError(err.as_ref()));
+                        CfiModule {
+                            features: Default::default(),
+                            cfi_status: ObjectFileStatus::from(err.as_ref()),
+                            cfi_path: None,
+                            scanned: false,
+                        }
+                    }
+                };
+                (*code_id, cfi_module)
+            })
+            .collect();
+        Self { inner }
+    }
+
+    /// Load the CFI information from the cache.
+    ///
+    /// This reads the CFI caches from disk and returns them in a format suitable for the
+    /// breakpad processor to stackwalk.  Loading the caches from disk may update the
+    /// [`CfiModule::cfi_status`].
+    fn load_cfi(&mut self) -> BTreeMap<CodeModuleId, CfiCache> {
+        self.inner
+            .iter_mut()
+            .filter_map(|(code_id, cfi_module)| {
+                let path = cfi_module.cfi_path.as_ref()?;
+                let bytes = ByteView::open(path)
+                    .map_err(|err| {
+                        log::error!("Error while reading cficache: {}", LogError(&err));
+                        cfi_module.cfi_status = ObjectFileStatus::Other;
+                        err
+                    })
+                    .ok()?;
+                let cfi_cache = CfiCache::from_bytes(bytes)
+                    .map_err(|err| {
+                        // This mostly never happens since we already checked the files
+                        // after downloading and they would have been tagged with
+                        // CacheStatus::Malformed.
+                        log::error!("Error while loading cficache: {}", LogError(&err));
+                        cfi_module.cfi_status = ObjectFileStatus::Other;
+                        err
+                    })
+                    .ok()?;
+                Some((*code_id, cfi_cache))
+            })
+            .collect()
+    }
+
+    /// Marks a module as scanned for CFI.
+    ///
+    /// If during stack unwinding the module was scanned that means we needed its CFI.  We
+    /// need to keep track of this because it may indicate we didn't fetch a CFI file we
+    /// needed, or we thought a CFI module was missing but it wasn't needed.  So this also
+    /// updates the [`CfiModule::cfi_status`] field.
+    fn mark_scanned(&mut self, code_id: CodeModuleId) {
+        match self.inner.get_mut(&code_id) {
```

Comment:
does `BTreeMap` have the same `entry` api as hashmaps? Maybe `entry().or_insert_with()` may be simpler?

Reply:
Yes it does.  I thought I didn't like it but tried that again and it is indeed nicer.